examples/quantize/quantize.cpp

```diff
@@ -4,43 +4,137 @@
 
 #include <cstdio>
 #include <cstring>
-#include <map>
+#include <vector>
+#include <iostream>
+#include <iomanip>
 #include <string>
 
-static const std::map<std::string, llama_ftype> LLAMA_FTYPE_MAP = {
-  {"q4_0",   LLAMA_FTYPE_MOSTLY_Q4_0},
-  {"q4_1",   LLAMA_FTYPE_MOSTLY_Q4_1},
-  {"q5_0",   LLAMA_FTYPE_MOSTLY_Q5_0},
-  {"q5_1",   LLAMA_FTYPE_MOSTLY_Q5_1},
-  {"q8_0",   LLAMA_FTYPE_MOSTLY_Q8_0},
-  {"q2_K",   LLAMA_FTYPE_MOSTLY_Q2_K},
-  {"q3_K",   LLAMA_FTYPE_MOSTLY_Q3_K_M},
-  {"q3_K_S", LLAMA_FTYPE_MOSTLY_Q3_K_S},
-  {"q3_K_M", LLAMA_FTYPE_MOSTLY_Q3_K_M},
-  {"q3_K_L", LLAMA_FTYPE_MOSTLY_Q3_K_L},
-  {"q4_K",   LLAMA_FTYPE_MOSTLY_Q4_K_M},
-  {"q4_K_S", LLAMA_FTYPE_MOSTLY_Q4_K_S},
-  {"q4_K_M", LLAMA_FTYPE_MOSTLY_Q4_K_M},
-  {"q5_K",   LLAMA_FTYPE_MOSTLY_Q5_K_M},
-  {"q5_K_S", LLAMA_FTYPE_MOSTLY_Q5_K_S},
-  {"q5_K_M", LLAMA_FTYPE_MOSTLY_Q5_K_M},
-  {"q6_K",   LLAMA_FTYPE_MOSTLY_Q6_K},
+struct quant_option {
+    std::string name;
+    llama_ftype ftype;
+    std::string desc;
 };
 
-bool try_parse_ftype(const std::string & ftype_str, llama_ftype & ftype, std::string & ftype_str_out) {
-    auto it = LLAMA_FTYPE_MAP.find(ftype_str);
-    if (it != LLAMA_FTYPE_MAP.end()) {
-        ftype = it->second;
-        ftype_str_out = it->first;
-        return true;
+static const std::vector<struct quant_option> QUANT_OPTIONS = {
+    {
+        "q4_0",
+        LLAMA_FTYPE_MOSTLY_Q4_0,
+        "approx +0.2499 perplexity, 3.50G output @ 7B",
```

Comment:
Just in case it wasn't clear, `+blah perplexity` means relative to the full quality model. Or maybe you were talking about how different sizes of models start at different perplexity values?

I agree that the `+x perplexity` thing isn't ideal but I'm not really sure of a better way to briefly give the user an idea of how that quantization level may affect the model quality. Do you have a better idea?

(Also, no problem with your suggested size output change.)

Reply:
>Or maybe you were talking about how different sizes of models start at different perplexity values?

That is what I meant.

I think (but definitely don't quote me on this) that perplexity is logarithmic. So if a model is one abstract unit of quality worse than another model the perplexity would go up by a fixed factor. If that is indeed the case then something like "+ X% perplexity" would be better and I would expect the relative change to be the same for different model sizes.

Reply:
I actually did a little research on this already:

## 7B

| name | +ppl | +ppl % | +ppl 13b to 7b % | size | size 16bit % | +ppl per -1G |
|-|-|-|-|-|-|-|
| q2_k     | 0.8698 | 14.726% | 133.344% | 2.67GB | 20.54% | 0.084201 |
| q3_ks    | 0.5505 | 9.320% | 84.394% | 2.75GB | 21.15% | 0.053707 |
| q3_km    | 0.2437 | 4.126% | 37.360% | 3.06GB | 23.54% | 0.024517 |
| q3_kl    | 0.1803 | 3.053% | 27.641% | 3.35GB | 25.77% | 0.018684 |
| q4_0     | 0.2499 | 4.231% | 38.311% | 3.50GB | 26.92% | 0.026305 |
| q4_1     | 0.1846 | 3.125% | 28.300% | 3.90GB | 30.00% | 0.020286 |
| q4_ks    | 0.1149 | 1.945% | 17.615% | 3.56GB | 27.38% | 0.012172 |
| q4_km    | 0.0535 | 0.906% | 8.202% | 3.80GB | 29.23% | 0.005815 |
| q5_0     | 0.0796 | 1.348% | 12.203% | 4.30GB | 33.08% | 0.009149 |
| q5_1     | 0.0415 | 0.703% | 6.362% | 4.70GB | 36.15% | 0.005000 |
| q5_ks    | 0.0353 | 0.598% | 5.412% | 4.33GB | 33.31% | 0.004072 |
| q5_km    | 0.0142 | 0.240% | 2.177% | 4.45GB | 34.23% | 0.001661 |
| q6_k     | 0.0044 | 0.074% | 0.675% | 5.15GB | 39.62% | 0.000561 |
| q8_0     | 0.0004 | 0.007% | 0.061% | 6.70GB | 51.54% | 0.000063 |


## 13B

| name | +ppl | +ppl % | +ppl 13b to 7b % | size | size 16bit % | +ppl per -1G |
|-|-|-|-|-|-|-|
| q2_k     | 0.6002 | 11.423% | 92.013% | 5.13GB | 20.52% | 0.030206 |
| q3_ks    | 0.3490 | 6.642% | 53.503% | 5.27GB | 21.08% | 0.017689 |
| q3_km    | 0.1955 | 3.721% | 29.971% | 5.88GB | 23.52% | 0.010225 |
| q3_kl    | 0.1520 | 2.893% | 23.302% | 6.45GB | 25.80% | 0.008194 |
| q4_0     | 0.1317 | 2.507% | 20.190% | 6.80GB | 27.20% | 0.007236 |
| q4_1     | 0.1065 | 2.027% | 16.327% | 7.60GB | 30.40% | 0.006121 |
| q4_ks    | 0.0861 | 1.639% | 13.199% | 6.80GB | 27.20% | 0.004731 |
| q4_km    | 0.0459 | 0.874% | 7.037% | 7.32GB | 29.28% | 0.002596 |
| q5_0     | 0.0313 | 0.596% | 4.798% | 8.30GB | 33.20% | 0.001874 |
| q5_1     | 0.0163 | 0.310% | 2.499% | 9.10GB | 36.40% | 0.001025 |
| q5_ks    | 0.0242 | 0.461% | 3.710% | 8.36GB | 33.44% | 0.001454 |
| q5_km    | 0.0095 | 0.181% | 1.456% | 8.60GB | 34.40% | 0.000579 |
| q6_k     | 0.0025 | 0.048% | 0.383% | 9.95GB | 39.80% | 0.000166 |
| q8_0     | 0.0005 | 0.010% | 0.077% | 13.00GB | 52.00% | 0.000042 |

***

So you're right, the percentages are pretty close. Personally I think the `+ppl 13b to 7b %` is most useful for comparisons but it requires too much explanation: that takes the relative perplexity increase as a percentage of the perplexity increase going from a full quality 13b LLaMA to the 7b version.

What do you think of including both figures like:

    approx +0.2499 (4.231%) perplexity, 3.50 GB size @ 7B

Reply:
Thank you for sharing those numbers. I think that the primary problem is that perplexity is a very abstract metric in the first place. It seems to me that both +ppl and +ppl% are not very good measures for how quantization affects results; providing two such numbers will only increase confusion. Maybe instead of just listing perplexity changes we can just add a print that says that a bigger model produces better results than a smaller model?

Reply:
> Maybe instead of just listing perplexity changes we can just add a print that says that a bigger model produces better results than a smaller model?

Unfortunately, that's not necessarily true. Consider the 7B `Q5_0` vs `Q4_K_M`. If they followed this advice, they'd choose `Q5_0` thinking it was better but actually end up with a bigger file and worse quality.

Another problem is that really gives no idea of how much of a difference there is. `F16` (or even `Q8_0`) is really big but you gain very little over `Q6_K`.

I'd generally rather give people a little information even if it's not ideal and then let them figure it out. 

> It seems to me that both +ppl and +ppl% are not very good measures for how quantization affects results;

That's pretty much why I like the comparing to 13B to 7B: people have an idea of how going from a 7B to a 13B feels, so putting the change in those terms gives them something to relate to. However, it would be complicated to explain in a simple tool like `quantize`.

It may not be worth really agonizing over this too much considering up until now there was _no_ information for the quantize types. More detailed explanations probably belong in the actual documentation rather than a brief list of quantization types.

Unless you have a better idea, I think it's probably a choice between just `+ppl`, just `+ppl%` or both like I mentioned before. Although it may be possible to also just add a brief comment, like saying `q4_k_m` is a good balance between size/quality, that kind of thing. Not sure I'm really qualified to write an evaluation of how good the quantization types are and which ones users should prefer though.